src/Gravity_level.hpp

```diff
@@ -0,0 +1,254 @@
+		}
+
+		// Do the (real) level solve (since in general the density has changed over
+		// the timestep due to hydro)
+
+		int is_new = 1;
+
+		solve_for_phi(level, phi_new, amrex::GetVecOfPtrs(get_grad_phi_curr(level)), is_new);
+
+		// [When level == sim->finestLevel(), the composite correction is zero, so
+		// only compute it for lower levels, if they exist.]
+
+		if (DoCompositeCorrection() != 0 && level < sim->finestLevel() && level <= get_max_solve_level()) {
+
+			// Add back the (composite - level) contribution.
+			phi_new.plus(*comp_minus_level_phi, 0, 1, 0);
+
+			for (int n = 0; n < AMREX_SPACEDIM; ++n) {
+				get_grad_phi_curr(level)[n]->plus(*comp_minus_level_grad_phi[n], 0, 1, 0);
+			}
+
+			if (test_results_of_solves() == 1) {
+				if (gravity::verbose != 0) {
+					amrex::Print() << "... testing grad_phi_curr (with composite correction)\n";
+				}
+				test_level_grad_phi_curr(level);
+			}
+		}
+	}
+
+	// Set new-time cell-centered gravitational acceleration vector
+	get_new_grav_vector(level, grav_new, time);
+
+	if (get_gravity_type() == GravityMode::Poisson && level < sim->finestLevel() && level <= get_max_solve_level()) {
+
+		if (DoCompositeCorrection() == 1) {
+			// Clear the pointers for the correction MultiFabs (deallocated via
+			// unique_ptr)
+			comp_minus_level_phi.reset();
+			comp_minus_level_grad_phi.clear();
+		}
+	}
+}
+
+template <typename T>
+void Gravity<T>::create_comp_minus_level_grad_phi(int level, MultiFab &comp_phi, const amrex::Vector<MultiFab *> &comp_gphi,
+						  std::unique_ptr<MultiFab> &comp_minus_level_phi,
+						  amrex::Vector<std::unique_ptr<MultiFab>> &comp_minus_level_grad_phi)
+{
+	BL_PROFILE("Gravity<T>::create_comp_minus_level_grad_phi()");
+
+	if (gravity::verbose > 1) {
+		amrex::Print() << "... saving the difference between composite and level solves\n";
+	}
+
+	// create comp_minus_level_phi
+
+	comp_minus_level_phi = std::make_unique<MultiFab>(sim->boxArray(level), sim->DistributionMap(level), 1, 0);
+
+	MultiFab::Copy(*comp_minus_level_phi, comp_phi, 0, 0, 1, 0);
+	comp_minus_level_phi->minus(phi_old_[level], 0, 1, 0);
+
+	// create comp_minus_level_grad_phi
+
+	comp_minus_level_grad_phi.resize(AMREX_SPACEDIM);
+
+	for (int n = 0; n < AMREX_SPACEDIM; ++n) {
+		comp_minus_level_grad_phi[n] =
+		    std::make_unique<MultiFab>(amrex::convert(sim->boxArray(level), IntVect::TheDimensionVector(n)), sim->DistributionMap(level), 1, 0);
+
+		MultiFab::Copy(*comp_minus_level_grad_phi[n], *comp_gphi[n], 0, 0, 1, 0);
+		comp_minus_level_grad_phi[n]->minus(*grad_phi_prev[level][n], 0, 1, 0);
+	}
+}
+
+template <typename T> void Gravity<T>::get_old_grav_vector(int level, MultiFab &grav_vector, Real time)
+{
+	BL_PROFILE("Gravity<T>::get_old_grav_vector()");
+
+	// Fill data from the level below if we're not doing a solve on this level
+
+	if (level > gravity::max_solve_level) {
+		sim->FillCoarsePatch(level, time, grav_vector, g_old_, g_new_, 0, 3, quokka::direction::na);
```

Comment:
I think we dont need `g_old_` and `g_new_` here anymore? This function expects integers as arguments at the 4th and 5th positions, and I think these integer arguments are 0 and 3.